### ISSUE TITLE
Native BFD support

### ISSUE THREAD

**Issue by Slepwin** [osrg/gobgp#2563]
Hi GoBGP team,

Thanks for your great progress on developing GoBGP.

Do you have any plans about implementing BFD natively in GoBGP? 
Direct and multihop sessions may be quite useful over non so reliable channels. Because BGP can't provide fast (ms) failure detection.

**Comment by fujita** [comment]
If someone sends a pull request for it, I would be happy to merge.

**Comment by dylandreimerink** [comment]
I have been working on this issue for a bit, [here](https://github.com/dylandreimerink/gobgp/compare/feature/bfd-support?expand=1) is what I have so far.

The state as of writing this is that I have basic BFD support, I can establish sessions with FRR on a single-hop connection, but no echo, multihop or authentication (all of which are optional BTW, so this is technically a minimal implementation of the spec).

I extended the gRPC API and hooked everything up on the server side. I have made two flags in the `global` and `neighbor add|update` commands to enable BFD with default settings as a demonstration. But I have doubts on what the best way is to add all of the BFD configuration and state info to the CLI. Would love some feedback on this @fujita if you can spare the time.

For the settings, the naive approach would be to add all settings to the existing param list:
```
usage: gobgp neighbor add [ <neighbor-address> | interface <neighbor-interface> ] as <VALUE> [ local-as <VALUE> | family <address-families-list> | vrf <vrf-name> | route-reflector-client [<cluster-id>] | route-server-client | allow-own-as <num> | remove-private-as (all|replace) | replace-peer-as | ebgp-multihop-ttl <ttl> | enable-bfd | bfd-passive | bfd-detect-multi <1-255> | bfd-tx-interval <ms> | bfd-rx-interval <ms> | bfd-demand | bfd-echo | bfd-echo-tx-interval <ms> | bfd-echo-rx-interval <ms>]
```
Given the amount of sessions, this seems like it would negatively impact the user experience. 

A second approach could be to define a separate BFD command.

Again, I can go two ways in this. The first would be to force users to define and use a "profile".
```
usage: gobgp bfd profile add <name> [passive | detect-multi <1-255> | tx-interval <ms> | rx-interval <ms> | demand | echo | echo-tx-interval <ms> | echo-rx-interval <ms>]
```

And then when enabling BFD for a neighbor, a user can specify a profile if they want a non default config.
```
usage: gobgp neighbor add [ <neighbor-address> | interface <neighbor-interface> ] as <VALUE> [ local-as <VALUE> | family <address-families-list> | vrf <vrf-name> | route-reflector-client [<cluster-id>] | route-server-client | allow-own-as <num> | remove-private-as (all|replace) | replace-peer-as | ebgp-multihop-ttl <ttl> | enable-bfd | bfd-profile <name> ]
```

Alternatively, we decouple BFD sessions completely from the BGP neighbors, and we allow users to optionally link a given BFD session to a BGP neighbor.

```
usage: gobgp bfd session add <name> <remote> [multihop | max-ttl <1-254> | passive | detect-multi <1-255> | tx-interval <ms> | rx-interval <ms> | demand | echo | echo-tx-interval <ms> | echo-rx-interval <ms>]
```

```
usage: gobgp neighbor add [ <neighbor-address> | interface <neighbor-interface> ] as <VALUE> [ local-as <VALUE> | family <address-families-list> | vrf <vrf-name> | route-reflector-client [<cluster-id>] | route-server-client | allow-own-as <num> | remove-private-as (all|replace) | replace-peer-as | ebgp-multihop-ttl <ttl> | bfd-session <name> ]
```

would really appreciate some thoughts on this.

**Comment by dabernie** [comment]
@dylandreimerink thanks for the update, I personally like the global BFD profiles definition and enabling BFD on a per BGP session with optional timer definitions at the peer level super-seeding the global profile

**Comment by YafimK** [comment]
@dylandreimerink I can't open the PR for some reason.
I'd be happy to contribute as needed to this PR... I like the second approach as well.

**Comment by fujita** [comment]
Out of curiosity, why integrating BFD into GoBGP is preferable? Having a separate BFD daemon that manages GoBGP via the gRPC APIs is simpler?

**Comment by dylandreimerink** [comment]
> Having a separate BFD daemon that manages GoBGP via the gRPC APIs is simpler?

Turns out it is, for a few reasons. BFD is all about the actual forwarding, so if you are running a route server, you want the actual BFD daemon to run on the devices that do the actual forwarding, not the control plane. If the BFD daemons detect trouble they contact the BGP daemon to perform some action.

Since BFD is routing protocol agnostic, it can also work with non BGP protocols like IS-IS and OSPF or even things like loadbalancers. 

Its for these two reasons that I decided to not implement in GoBGP directly

**Comment by fujita** [comment]
Yeah, makes sense. In the past, I implemented BFD in the same way. If your BFD implementation needs a new gRPC API to manage GoBGP, feel free to make a pull request.

**Comment by maxpain** [comment]
Any updates on this?

**Comment by dantte-lp** [comment]
For people landing on this thread now: the "separate BFD daemon over gRPC" path that fujita and dylandreimerink discussed upthread is what GoBFD does today.

https://github.com/dantte-lp/gobfd

`v0.6.1` ships RFC 5880/5881/5882/5883/7419/9384/9468/9747/9764, with the GoBGP integration calling `DisablePeer`/`EnablePeer` on BFD state changes (RFC 5882 §3.2 flap dampening). Runnable demo at `deployments/integrations/bgp-fast-failover/`.

Echo (RFC 9747) and Micro-BFD (RFC 7130) live behind their own gRPC services because their timer / aggregation models do not fit the base `AddSession` shape.

#3388 takes the in-process route; both have their place. If a small gRPC verb on GoBGP would help the standalone path (e.g. richer state-change notifications during graceful restart), I can draft a PR.

Disclosure: I'm the author.

**Comment by fujita** [comment]
GoBGP should clearly focus on BGP implementation. If the approach of integrating with an external BFD daemon via gRPC works, I'd prefer to go in that direction.

@Ivan-Pokhabov opinions?

**Comment by Ivan-Pokhabov** [comment]
> GoBGP should clearly focus on BGP implementation. If the approach of integrating with an external BFD daemon via gRPC works, I'd prefer to go in that direction.
> 
> [@Ivan-Pokhabov](https://github.com/Ivan-Pokhabov) opinions?

I see the value in keeping GoBGP focused, but I'd like to propose merging #3388 as a lightweight, built-in solution. It covers standard use cases and provides a out-of-the-box solution — it can be enabled with just a couple of lines in the config without the operational overhead of running a separate daemon.

We could use a hybrid approach: provide this simple built-in BFD for default cases, but fully support integrating with external daemons (like GoBFD) via gRPC for users who need advanced functionality and full RFC compliance.

**Comment by fujita** [comment]
@dantte-lp Could you share how mature GoBFD currently is? Is it already being used in production environments?

If GoBFD is stable and production-ready, I don't think GoBGP needs to bundle its own BFD implementation — users who need BFD can simply integrate with GoBFD via the gRPC approach.

**Comment by Ivan-Pokhabov** [comment]
@fujita I'd like to provide some additional context and outline the reasons why #3388 would be beneficial for the GoBGP ecosystem:

1. Maturity and strict RFC compliance.
This is a complete implementation of RFC 5880 and 5881. It has been running successfully in our production environment under real-world workloads for 2 years, interoperating seamlessly with other BFD implementations.

2. Strong demand from Cloud-Native projects.
The lack of a built-in BFD is a known pain point for major systems relying on GoBGP. Communities like Cilium, Talos, and kube-router have actively requested this functionality. In Kubernetes environments, deploying and managing a separate standalone BFD daemon (e.g., as a sidecar) introduces significant operational overhead. A built-in solution solves this elegantly.

- https://github.com/siderolabs/talos/issues/11126
- https://github.com/cloudnativelabs/kube-router/issues/1117
- https://github.com/cilium/cilium/issues/22394
- https://github.com/runityru/anycastd/issues/88

3. Feature parity.
GoBGP's main alternatives, such as FRR and BIRD, provide built-in BFD support out-of-the-box. Users and network engineers naturally expect this baseline functionality from any modern routing daemon.

4. No conflict with external daemons.
I want to emphasize that the built-in implementation is entirely optional (toggleable). It does not prevent or interfere with the use of external solutions. We are simply giving users a choice: they can use the built-in, zero-dependency BFD for standard use cases, or disable it and use an external daemon (like GoBFD) via gRPC for advanced scenarios.

Finally, the code footprint of this PR is small, so it won't create a maintenance burden. This establishes a clear operational boundary: GoBGP provides a reliable, minimal built-in BFD, and we can deliberately delegate any future requests for complex BFD features to external standalone daemons.

**Comment by fujita** [comment]
I've just merged built-in BFD support https://github.com/osrg/gobgp/pull/3402

Personally, I still lean toward the external daemon approach — in production environments, GoBGP is typically deployed alongside many other software components anyway, so I'm not convinced that moving BFD in-process makes a substantial difference. That said, there seems to be clear demand for the built-in approach, and there's no mature external BFD implementation to point people to. Let's see how it works out.

**Comment by fujita** [comment]
Closing this now that built-in BFD support has been added https://github.com/osrg/gobgp/pull/3402
